source/parsed_preconditioner.cc

```diff
@@ -0,0 +1,276 @@
+                "set to smoother_sweeps. The term sweeps refers to the number of\n"
+                "matrix-vector products performed in the Chebyshev case. In the\n"
+                "non-elliptic case, smoother_sweeps sets the number of SSOR relaxation\n"
+                "sweeps for post-smoothing to be performed.");
+
+  add_parameter(prm, &smoother_overlap, "Smoother overlap", std::to_string(smoother_overlap),
+                Patterns::Integer(0),
+                "Determines the overlap in the SSOR/Chebyshev error smoother when run\n"
+                "in parallel.");
+
+  add_parameter(prm, &output_details, "Output details", output_details ? "true":"false",
+                Patterns::Bool(),
+                "If this flag is set to true, then internal information from the ML\n"
+                "preconditioner is printed to screen. This can be useful when debugging\n"
+                "the preconditioner.");
+
+  add_parameter(prm, &smoother_type, "Smoother type", smoother_type,
+                Patterns::Selection("|Aztec|IFPACK|Jacobi"
+                                    "|ML symmetric Gauss-Seidel|symmetric Gauss-Seidel"
+                                    "|ML Gauss-Seidel|Gauss-Seidel|block Gauss-Seidel"
+                                    "|symmetric block Gauss-Seidel|Chebyshev|MLS|Hiptmair"
+                                    "|Amesos-KLU|Amesos-Superlu|Amesos-UMFPACK|Amesos-Superludist"
+                                    "|Amesos-MUMPS|user-defined|SuperLU|IFPACK-Chebyshev|self"
+                                    "|do-nothing|IC|ICT|ILU|ILUT|Block Chebyshev"
+                                    "|IFPACK-Block Chebyshev"),
+                "Determines which smoother to use for the AMG cycle.");
+
+  add_parameter(prm, &coarse_type, "Coarse type", coarse_type,
+                Patterns::Selection("|Aztec|IFPACK|Jacobi"
+                                    "|ML symmetric Gauss-Seidel|symmetric Gauss-Seidel"
+                                    "|ML Gauss-Seidel|Gauss-Seidel|block Gauss-Seidel"
+                                    "|symmetric block Gauss-Seidel|Chebyshev|MLS|Hiptmair"
+                                    "|Amesos-KLU|Amesos-Superlu|Amesos-UMFPACK|Amesos-Superludist"
+                                    "|Amesos-MUMPS|user-defined|SuperLU|IFPACK-Chebyshev|self"
+                                    "|do-nothing|IC|ICT|ILU|ILUT|Block Chebyshev"
+                                    "|IFPACK-Block Chebyshev"),
+                "Determines which solver to use on the coarsest level. The same\n"
+                "settings as for the smoother type are possible.");
+}
+
+template<int dim, int spacedim, typename Matrix>
+void ParsedAMGPreconditioner::initialize_preconditioner( const ParsedFiniteElement<dim, spacedim> &fe,
+                                                         const DoFHandler<dim, spacedim> &dh,
+                                                         const Matrix &matrix)
```

Comment:
`typename std::enable_if<(dim==spacedim), void **>::type=0` such that we can instantiate the class for all cases we deal with in pi-DoMUS